File: packages/fluse_protocol/lib/src/tunnel_frame.dart
import 'dart:typed_data';

import 'protocol_exception.dart';

/// トンネルフレームの種別（設計 §2.2.1）。
enum TunnelOpcode {
  /// 新しいストリームを開く。
  open(0x01),

  /// データ本体。
  data(0x02),

  /// ストリームを閉じる。
  close(0x03);

  const TunnelOpcode(this.value);

  /// バイト0に載る値。
  final int value;

  static TunnelOpcode? tryParse(int value) {
    for (final TunnelOpcode opcode in values) {
      if (opcode.value == value) {
        return opcode;
      }
    }
    return null;
  }
}

/// WebSocket の binary frame に載せる TCP トンネルのフレーム
/// （設計 §2.2.1）。
///
/// ```text
/// byte0      : opcode  0x01=open, 0x02=data, 0x03=close
/// byte1..4   : streamId (uint32 big-endian)
/// byte5..    : payload (data時のみ)
/// ```
///
/// **VM Service のプロトコルは一切解釈しない**（設計 §10-3）。
/// JSON-RPC over WebSocket と DevFS の HTTP PUT が同じポートに来るため、
/// 片方だけ対応した「賢い」中継を書くと必ず破綻する。ここは生の
/// バイト列を運ぶだけに徹する。
final class TunnelFrame {
  const TunnelFrame({
    required this.opcode,
    required this.streamId,
    this.payload = _empty,
  });

  /// ストリームを開くフレーム。
  TunnelFrame.open(this.streamId)
    : opcode = TunnelOpcode.open,
      payload = _empty;

  /// データを運ぶフレーム。
  TunnelFrame.data(this.streamId, this.payload) : opcode = TunnelOpcode.data;

  /// ストリームを閉じるフレーム。
  TunnelFrame.close(this.streamId)
    : opcode = TunnelOpcode.close,
      payload = _empty;

  static const List<int> _empty = <int>[];

  /// ヘッダの長さ（opcode 1バイト + streamId 4バイト）。
  static const int headerLength = 5;

  /// `streamId` の最大値。uint32 なので 0xFFFFFFFF。
  static const int maxStreamId = 0xFFFFFFFF;

  /// 1フレームに載せられる payload の上限（1 MiB）。
  ///
  /// **送信側はこれを超える前に分割する責務がある。**
  /// 上限を設けないと、壊れた相手や悪意ある相手が送る長さ宣言で
  /// メモリを一気に確保させられる。設計 §8.2-5 のバックプレッシャ閾値
  /// （送信キュー 4MB）より十分小さくし、1フレームでキューを埋めない
  /// 大きさにしてある。
  static const int maxPayloadLength = 1024 * 1024;

  final TunnelOpcode opcode;

  /// ストリームの識別子。uint32。
  final int streamId;

  /// 本体。`data` 以外では空。
  final List<int> payload;

  Uint8List encode() {
    if (streamId < 0 || streamId > maxStreamId) {
      throw FluseProtocolException('streamId が uint32 の範囲外です: $streamId');
    }
    if (opcode != TunnelOpcode.data && payload.isNotEmpty) {
      // open / close に本体を載せると、受け側の解釈が opcode と食い違う。
      throw FluseProtocolException(
        '${opcode.name} フレームに payload は載せられません'
        '（${payload.length} バイト）',
      );
    }
    if (payload.length > maxPayloadLength) {
      throw FluseProtocolException(
        'payload が上限を超えています: ${payload.length} バイト'
        '（上限 $maxPayloadLength）。送信側で分割してください',
      );
    }
    for (int i = 0; i < payload.length; i++) {
      final int byte = payload[i];
      if (byte < 0 || byte > 0xFF) {
        // Uint8List.setRange は範囲外を例外にせず下位8ビットへ切り詰める。
        // 黙って別のバイト列を送ることになる。
        throw FluseProtocolException('payload の $i 番目がバイトの範囲外です: $byte');
      }
    }

    final Uint8List bytes = Uint8List(headerLength + payload.length);
    bytes[0] = opcode.value;
    // big-endian。バイト順を取り違えると streamId が別のストリームを指す。
    bytes[1] = (streamId >> 24) & 0xFF;
    bytes[2] = (streamId >> 16) & 0xFF;
    bytes[3] = (streamId >> 8) & 0xFF;
    bytes[4] = streamId & 0xFF;
    bytes.setRange(headerLength, bytes.length, payload);
    return bytes;
  }

  static TunnelFrame decode(List<int> bytes) {
    if (bytes.length < headerLength) {
      throw FluseProtocolException(
        'トンネルフレームが短すぎます: ${bytes.length} バイト'
        '（最低 $headerLength バイト必要）',
      );
    }

    // ヘッダを読む前に全要素を検査する。範囲外が混ざっていると、
    // streamId が別の値になり、payload は Uint8List.fromList で黙って
    // 下位8ビットへ切り詰められる。どちらも気づけない壊れ方をする。
    for (int i = 0; i < bytes.length; i++) {
      final int byte = bytes[i];
      if (byte < 0 || byte > 0xFF) {
        throw FluseProtocolException('フレームの $i 番目がバイトの範囲外です: $byte');
      }
    }

    final TunnelOpcode? opcode = TunnelOpcode.tryParse(bytes[0]);
    if (opcode == null) {
      // 未知の opcode を黙って捨てると、相手は届いたと思って待ち続ける。
      throw FluseProtocolException(
        '未知の opcode: 0x${bytes[0].toRadixString(16).padLeft(2, '0')}',
      );
    }

    final int streamId =
        (bytes[1] << 24) | (bytes[2] << 16) | (bytes[3] << 8) | bytes[4];

    final int payloadLength = bytes.length - headerLength;
    if (payloadLength > maxPayloadLength) {
      // コピーする前に弾く。長さを信じて確保すると、壊れた相手に
      // メモリを一気に取らせられる。
      throw FluseProtocolException(
        'payload が上限を超えています: $payloadLength バイト'
        '（上限 $maxPayloadLength）',
      );
    }

    final List<int> payload = payloadLength == 0
        ? _empty
        : Uint8List.fromList(bytes.sublist(headerLength));

    if (opcode != TunnelOpcode.data && payload.isNotEmpty) {
      throw FluseProtocolException(
        '${opcode.name} フレームに payload が付いています（${payload.length} バイト）',
      );
    }

    return TunnelFrame(opcode: opcode, streamId: streamId, payload: payload);
  }

  @override
  String toString() =>
      'TunnelFrame(${opcode.name}, stream: $streamId, '
      '${payload.length}バイト)';
}

File: packages/fluse_runtime/android/src/wire/kotlin/dev/fluse/protocol/TunnelFrame.kt
package dev.fluse.protocol

/** トンネルフレームの種別（設計 §2.2.1）。 */
enum class TunnelOpcode(val value: Int) {
    /** 新しいストリームを開く。 */
    OPEN(0x01),

    /** データ本体。 */
    DATA(0x02),

    /** ストリームを閉じる。 */
    CLOSE(0x03),

    ;

    /** Dart 側の enum 名と揃えたワイヤ表現（ゴールデンの `opcode`）。 */
    val wireName: String get() = name.lowercase()

    companion object {
        fun tryParse(value: Int): TunnelOpcode? = entries.firstOrNull { it.value == value }

        fun tryParseName(name: String): TunnelOpcode? =
            entries.firstOrNull { it.wireName == name }
    }
}

/**
 * WebSocket の binary frame に載せる TCP トンネルのフレーム（設計 §2.2.1）。
 *
 * ```text
 * byte0      : opcode  0x01=open, 0x02=data, 0x03=close
 * byte1..4   : streamId (uint32 big-endian)
 * byte5..    : payload (data時のみ)
 * ```
 *
 * **VM Service のプロトコルは一切解釈しない**（設計 §10-3）。
 *
 * Dart 側の `TunnelFrame` と完全に同じ表現でなければならない。
 * 検証は `packages/fluse_protocol/test/fixtures/wire_golden.json` を
 * 両実装が読むことで担保する。
 */
class TunnelFrame(
    val opcode: TunnelOpcode,
    /** ストリームの識別子。uint32 なので Kotlin では [Long] で持つ。 */
    val streamId: Long,
    /** 本体。`data` 以外では空。 */
    val payload: ByteArray = EMPTY,
) {
    companion object {
        private val EMPTY = ByteArray(0)

        /** ヘッダの長さ（opcode 1バイト + streamId 4バイト）。 */
        const val HEADER_LENGTH = 5

        /** `streamId` の最大値。uint32 なので 0xFFFFFFFF。 */
        const val MAX_STREAM_ID = 0xFFFFFFFFL

        /**
         * 1フレームに載せられる payload の上限（1 MiB）。
         *
         * 送信側はこれを超える前に分割する責務がある。
         */
        const val MAX_PAYLOAD_LENGTH = 1024 * 1024

        /** ストリームを開くフレーム。 */
        fun open(streamId: Long): TunnelFrame = TunnelFrame(TunnelOpcode.OPEN, streamId)

        /** データを運ぶフレーム。 */
        fun data(streamId: Long, payload: ByteArray): TunnelFrame =
            TunnelFrame(TunnelOpcode.DATA, streamId, payload)

        /** ストリームを閉じるフレーム。 */
        fun close(streamId: Long): TunnelFrame = TunnelFrame(TunnelOpcode.CLOSE, streamId)

        fun decode(bytes: ByteArray): TunnelFrame {
            if (bytes.size < HEADER_LENGTH) {
                throw FluseProtocolException(
                    "トンネルフレームが短すぎます: ${bytes.size} バイト（最低 $HEADER_LENGTH バイト必要）",
                )
            }

            val opcode = TunnelOpcode.tryParse(bytes[0].toInt() and 0xFF)
                ?: throw FluseProtocolException(
                    "未知の opcode: 0x%02x".format(bytes[0].toInt() and 0xFF),
                )

            val payloadLength = bytes.size - HEADER_LENGTH
            if (payloadLength > MAX_PAYLOAD_LENGTH) {
                // コピーする前に弾く。長さを信じて確保すると、壊れた相手に
                // メモリを一気に取らせられる。
                throw FluseProtocolException(
                    "payload が上限を超えています: $payloadLength バイト（上限 $MAX_PAYLOAD_LENGTH）",
                )
            }

            // Kotlin の Byte は符号付き。マスクせずに組み立てると
            // streamId が負になり、別のストリームを指す。
            val streamId =
                ((bytes[1].toLong() and 0xFF) shl 24) or
                    ((bytes[2].toLong() and 0xFF) shl 16) or
                    ((bytes[3].toLong() and 0xFF) shl 8) or
                    (bytes[4].toLong() and 0xFF)

            val payload =
                if (payloadLength == 0) EMPTY else bytes.copyOfRange(HEADER_LENGTH, bytes.size)

            if (opcode != TunnelOpcode.DATA && payload.isNotEmpty()) {
                throw FluseProtocolException(
                    "${opcode.wireName} フレームに payload が付いています（${payload.size} バイト）",
                )
            }

            return TunnelFrame(opcode, streamId, payload)
        }
    }

    fun encode(): ByteArray {
        if (streamId < 0 || streamId > MAX_STREAM_ID) {
            throw FluseProtocolException("streamId が uint32 の範囲外です: $streamId")
        }
        if (opcode != TunnelOpcode.DATA && payload.isNotEmpty()) {
            // open / close に本体を載せると、受け側の解釈が opcode と食い違う。
            throw FluseProtocolException(
                "${opcode.wireName} フレームに payload は載せられません（${payload.size} バイト）",
            )
        }
        if (payload.size > MAX_PAYLOAD_LENGTH) {
            throw FluseProtocolException(
                "payload が上限を超えています: ${payload.size} バイト" +
                    "（上限 $MAX_PAYLOAD_LENGTH）。送信側で分割してください",
            )
        }

        val bytes = ByteArray(HEADER_LENGTH + payload.size)
        bytes[0] = opcode.value.toByte()
        // big-endian。バイト順を取り違えると streamId が別のストリームを指す。
        bytes[1] = ((streamId shr 24) and 0xFF).toByte()
        bytes[2] = ((streamId shr 16) and 0xFF).toByte()
        bytes[3] = ((streamId shr 8) and 0xFF).toByte()
        bytes[4] = (streamId and 0xFF).toByte()
        payload.copyInto(bytes, HEADER_LENGTH)
        return bytes
    }

    override fun toString(): String =
        "TunnelFrame(${opcode.wireName}, stream: $streamId, ${payload.size}バイト)"
}
